### DISCUSSION TITLE
Optimal microstrip lumped port size/placement and Thirds Rule question.

### DISCUSSION THREAD

**Discussion by georgy-moshkin** [thliebig/openEMS-Project#200] [Q&A Usage]
![openems-lumped-port-placement](https://github.com/thliebig/openEMS-Project/assets/48346176/ec80a900-734a-4381-955b-58dae2aa0f2b)
[FDTD Mesh](https://wiki.openems.de/index.php/FDTD_Mesh.html) page of OpenEMS wiki mentions Thirds Rule: "...to put a mesh line 1/3 inside and 2/3 outside the metal." From my understanding, the thirds rule is especially important for sparse FDTD grid (low resolution grid). Image above shows four arrangements:
1 - "standard" placement.  This is what we usually see in commercial simulators. Note that port width is not aligned with vertical mesh lines. I don't know if this a problem, but port will not work if it is not aligned with horizontal mesh line.
2 - Horizontal mesh line and lumped port are placed 1/3 * max_res from microstrip edge. Lumped port thickness is 2 * 3 * max_res.
3 - The same as (2), but port width is decreased to be aligned with "1/3" rule mesh lines
4 - The same as (1), but horizontal mesh line and port are placed 1/3 * max_res away from microstrip edge
I know that using 1/3 rule doesn't make much sense on a port (vertical direction), it's just to make grid and placement more uniform. 
My question is about port width. Should it be aligned with mesh lines or with metal width? I use OpenEMS for my educational project and want to use best approach in my examples. Most likely case 1 is the best option if mesh is dense enough. Are there any know-hows on port placement, especially for sparse mesh?

Case 1:
```
start = [-feed.width/2 feed.pos 0];
stop  = [feed.width/2 feed.pos substrate.thickness];

mesh.x = [mesh.x -feed.width/2-max_res/2*0.66 -feed.width/2+max_res/2*0.33 feed.width/2+max_res/2*0.66 feed.width/2-max_res/2*0.33]
mesh.y = [mesh.y feed.pos];
```

Case 2:
```
start = [-feed.width/2 feed.pos+2*max_res/2*0.33 0];
stop  = [feed.width/2 feed.pos substrate.thickness]

mesh.x = [mesh.x -feed.width/2-max_res/2*0.66 -feed.width/2+max_res/2*0.33 feed.width/2+max_res/2*0.66 feed.width/2-max_res/2*0.33]
mesh.y = [mesh.y feed.pos+max_res/2*0.33];
```

Case 3:
```
start = [-feed.width/2+max_res/2*0.33 feed.pos+2*max_res/2*0.33 0];
stop  = [feed.width/2-max_res/2*0.33 feed.pos substrate.thickness]

mesh.x = [mesh.x -feed.width/2-max_res/2*0.66 -feed.width/2+max_res/2*0.33 feed.width/2+max_res/2*0.66 feed.width/2-max_res/2*0.33]
mesh.y = [mesh.y feed.pos+max_res/2*0.33];
```

Case 4:
```
start = [-feed.width/2 feed.pos+max_res/2*0.33 0];
stop  = [feed.width/2 feed.pos+max_res/2*0.33 substrate.thickness];

mesh.x = [mesh.x -feed.width/2-max_res/2*0.66 -feed.width/2+max_res/2*0.33 feed.width/2+max_res/2*0.66 feed.width/2-max_res/2*0.33]
mesh.y = [mesh.y feed.pos+max_res/2*0.33];
```

**Comment by georgy-moshkin** [comment]
Answering my own question. I did a lot of tests using 50 Ohm line and came up to a conclusion that first case placement is enough. I found out that unwanted variation in impedance was caused by the automatically generated mesh and not by the port itself. I excluded port from DetectEdges() function:
mesh = DetectEdges(CSX, mesh,'ExcludeProperty',{'patch','network','port_excite_1','port_resist_1'});
and created mesh lines manually. I've noticed that 4 cells + 1/3 rule at the edges may not be enough: 50 Ohm impedance sometimes goes up to 60 Ohm, which causes a accumulating mismatch on power dividers connected in series. After adding more cells manually I get correct results. To sum up, I used the wrong approach: tried different port geometry while the problem was caused by meshing.

**Reply by Miceuz** [reply]
@georgy-moshkin I am going thru similar steep self-teaching curve. 

Can you please elaborate a bit more on 4+1/3 rule - do you mean 2 cells inside of microstrip and 2 on the sides where 1/3 of them is taken by microstrip, just like in your picture?

Then you state that this is not enough to achieve accuracy. Can you elaborate and maybe post a screenshot - what  is enough for accuracy in your case? Do you split internal 2 cells into more, say 4 and leave 2 coarse cells on the outside or do you split to finer mesh but still make sure that external cells still comply with 1/3 rule?

Basically my question is where do we need finer mesh - inside of microstrip only or both inside and on the boundary of microstrip?

**Reply by georgy-moshkin** [reply]
We need finer mesh on the edges (left, right) and possibly at the ports too. Or simply use finer mesh everywhere.
Here is what I did:
```
  % "1/3" rule for patches and inset cutouts

  for sx = 0:patch.NX-1
    offs.X=sx*patch.step-(patch.NX-1)*patch.step/2;
    x=-patch.width/2+offs.X;

    mesh.x = [mesh.x  x-max_res*0.66  x+max_res*0.33];
    mesh.x = [mesh.x -x+max_res*0.66 -x-max_res*0.33];

    x=feed.width/2+inset.width+offs.X;
    mesh.x = [mesh.x  x-max_res*0.66  x+max_res*0.33];
    mesh.x = [mesh.x -x+max_res*0.66 -x-max_res*0.33];
  endfor

  for sy = 0:patch.NY-1
    offs.Y=sy*patch.step-(patch.NY-1)*patch.step/2;
    y=-patch.length/2+offs.Y;
    mesh.y = [mesh.y  y-max_res*0.66  y+max_res*0.33];
    mesh.y = [mesh.y -y+max_res*0.66 -y-max_res*0.33];

    if (mod(sy, 2) != 0)
      y=-patch.length/2+inset.length+offs.Y;
      mesh.y = [mesh.y  y-max_res*0.66  y+max_res*0.33];
    else
      y=-(-patch.length/2+inset.length)+offs.Y;
      mesh.y = [mesh.y  y+max_res*0.66  y-max_res*0.33];
    endif

  endfor


  % Extra cells for main feeding network
  res=max_res/4;

  % "1/3" rule for 50 ohm feeding lines
  mesh.y = [mesh.y -mainFeed.width/2+res*0.33+move90 mainFeed.width/2-res*0.33+move90];
  mesh.y = [mesh.y -mainFeed.width/2-res*0.66+move90 mainFeed.width/2+res*0.66+move90];
  mesh.x = [mesh.x -mainFeed.width/2+res*0.33 mainFeed.width/2-res*0.33];
  mesh.x = [mesh.x -mainFeed.width/2-res*0.66 mainFeed.width/2+res*0.66];

  % "1/3" rule for quarterwave transformer
  mesh.x = [mesh.x -trans.W/2+res*0.33 trans.W/2-res*0.33];
  mesh.x = [mesh.x -trans.W/2-res*0.66 trans.W/2+res*0.66];
  mesh.y = [mesh.y mainFeed.width/2+trans.L-res*0.33+move90];
  mesh.y = [mesh.y mainFeed.width/2+trans.L+res*0.66+move90];

  % mesh lines for 100 Ohm sections
  mesh.x = [mesh.x linspace(-patch.step/2-feed.width/2,-patch.step/2+feed.width/2,4+1)];
  mesh.x = [mesh.x linspace(patch.step/2-feed.width/2,patch.step/2+feed.width/2,4+1)];

  % mesh lines for feeding port
  mesh.y = [mesh.y mainFeed.posY];
  mesh.y = [mesh.y mainFeed.posY+res];
  mesh.y = [mesh.y mainFeed.posY-res];


  mesh = SmoothMesh(mesh, max_res*2, 1.25, 'algorithm',3); % lgorithm=3 works nice

  CSX = DefineRectGrid(CSX, unit, mesh);
```

The most important part is that feeding network uses finer mesh than the patches (everything below "res=max_res/4" line).
Plus, I've added two more meshing lines near the feeding port. And then, final SmoothMesh(mesh, max_res*2,...
To sum up, 
1) patches with resolution max_res
2) feeding network with resolution max_res/4
3) smooth mesh with resolution max_res/2 (this is for decrease computation time only, max_res/4 will be better).

For example, https://wiki.openems.de/index.php/FDTD_Mesh.html shows
a great image, but resolution is too low. Works for education, but before manufacturing this either correct result by some correction coefficient or something like that to shift S11/S22 measurements. A simple solution is to use higher resolution mesh for everything, but then need to wait much longer. So, we can force a finer mesh by placing manual max_res/4 lines on the edges of the line, and then let SmoothMesh() smooth it to a lower resolution mesh in empty space.

**Reply by KajWiik** [reply]
This thread saves a lot of time for me, many thanks!

**Reply by georgy-moshkin** [reply]
Here is a more intuitive explanation based on computer graphics example. Following image demonstrates 3D scene direct and indirect light calculation (radiosity approximation algorithm https://www.researchgate.net/publication/239562773).

![image](https://github.com/user-attachments/assets/f8ba608e-e0ae-4556-b509-ebc423546e39)

Low-polygonal model on the right was used to calculate the light intensities, e.g., at the edges of each triangle. Then this low-resolution light distribution can be cleverly averaged to get the light distribution for the image on the left. Note that if we use less triangles for the low-resolution model, then it will be impossible to retrieve details such as shadow under the sphere. For example, if the floor consists only of two large triangles (4 vertices = 4 intensity values). The similar effect happens with FDTD: we can't get realistic electromagnetic light distribution if the polygonal mesh is too rough.

**Reply by KajWiik** [reply]
@georgy-moshkin I am still struggling with LumpedPorts, [they have very slow convergence related to MSLPorts](https://github.com/thliebig/openEMS-Project/discussions/323#discussion-8318030). What are your boundary sizes and conditions, it looks like that the ground plane is larger (longer) than the microstrip? Could you share your code?
